dynamodb/schema.go

```diff
@@ -0,0 +1,367 @@
+	}
+}
+
+type statItem struct {
+	Type  string
+	Count int64
+}
+
+func (s *dynamoDBSchema) inferDataTypes(stats map[string]map[string]int64, rows int64) {
+	if s.ColumnTypes == nil {
+		s.ColumnTypes = make(map[string]colType)
+	}
+
+	for col, countMap := range stats {
+		var statItems, candidates []statItem
+		var presentRows, normRows int64
+		for k, v := range countMap {
+			presentRows += v
+			if float64(v)/float64(rows) <= errThreshold {
+				// If the percentage is less than the error threshold, then
+				// this data type has a high chance to be mistakenly inserted
+				// and we should discard it.
+				continue
+			}
+			statItems = append(statItems, statItem{Type: k, Count: v})
+			normRows += v
+		}
+		if normRows == 0 {
+			log.Printf("Skip column %v with no data records", col)
+			continue
+		}
+
+		nullable := float64(rows-presentRows)/float64(rows) > errThreshold
+
+		for _, si := range statItems {
+			if float64(si.Count)/float64(normRows) > conflictThreshold {
+				// If the normalized percentage is greater than the conflicting
+				// threshold, we should consider this data type as a candidate.
+				candidates = append(candidates, si)
+			}
+		}
+
+		s.ColumnNames = append(s.ColumnNames, col)
+		if len(candidates) == 1 {
+			s.ColumnTypes[col] = colType{Name: candidates[0].Type, Nullable: nullable}
+		} else {
+			// If there is no any candidate or more than a single candidate,
+			// this column has a significant conflict on data types and then
+			// defaults to a String type.
+			s.ColumnTypes[col] = colType{Name: typeString, Nullable: nullable}
+		}
+	}
+}
+
+func (s *dynamoDBSchema) genericSchema() schema.Table {
+	colDefs := make(map[string]schema.Column)
+
+	for _, colName := range s.ColumnNames {
+		colType := s.ColumnTypes[colName]
+		colDef := schema.Column{
+			Name:    colName,
+			Type:    schema.Type{Name: colType.Name},
+			NotNull: !colType.Nullable,
+		}
+		colDefs[colName] = colDef
+	}
+
+	// Sort column names in increasing order.
+	colNames := make([]string, len(s.ColumnNames))
+	copy(colNames, s.ColumnNames)
+	sort.Strings(colNames)
+
+	// The order of primary keys is important.
+	var schemaPKeys []schema.Key
+	for _, colName := range s.PrimaryKeys {
+		schemaPKeys = append(schemaPKeys, schema.Key{Column: colName})
+		colDef := colDefs[colName]
+		colDefs[colName] = schema.Column{
+			Name:    colName,
+			Type:    colDef.Type,
+			NotNull: true,
+		}
+	}
+
+	// Record secondary indexes.
+	var indexes []schema.Index
+	for _, ind := range s.SecIndexes {
+		var keys []schema.Key
+		for _, k := range ind.Keys {
+			keys = append(keys, schema.Key{Column: k})
+		}
+		index := schema.Index{
+			Name: ind.Name,
+			Keys: keys,
+		}
+		indexes = append(indexes, index)
+	}
+
+	return schema.Table{
+		Name:        s.TableName,
+		ColNames:    colNames,
+		ColDefs:     colDefs,
+		PrimaryKeys: schemaPKeys,
+		Indexes:     indexes,
+	}
+}
+
+func numericParsable(n string) bool {
```

Comment:
This is a bit subtle.
I think it would be worth a function-level comment along the lines of:

numericParsable determines whether its argument is a valid
Spanner numeric values. This is based on <link to Spanner documentation of Spanner numeric datatype>.

and maybe a comment in the code about what the calculation is doing.